Source/Diagnostics/ReducedDiags/ParticleMeanEnergy.cpp

```diff
@@ -0,0 +1,155 @@
+#include <iostream>
+#include <cmath>
+
+/// constructor
+ParticleMeanEnergy::ParticleMeanEnergy (
+std::string rd_name, std::ofstream & ofs )
+: ReducedDiags{rd_name}
+{
+    /// get WarpX class object
+    auto & warpx = WarpX::GetInstance();
+
+    /// get MultiParticleContainer class object
+    auto & mypc = warpx.GetPartContainer();
+
+    /// get number of species (int)
+    auto species_number = mypc.nSpecies();
+
+    /// get species names (std::vector<std::string>)
+    auto species_names = mypc.GetSpeciesNames();
+
+    // write header row
+    ofs << "#";
+    ofs << "step";
+    ofs << m_sep;
+    ofs << "time(s)";
+    ofs << m_sep;
+    ofs << "total(J)";
+    for (int i = 0; i < species_number; ++i)
+    {
+        ofs << m_sep;
+        ofs << species_names[i]+"(J)";
+    }
+    ofs << std::endl;
+}
+///< end constructor
+
+/// destructor
+ParticleMeanEnergy::~ParticleMeanEnergy ()
+{}
+///< end destructor
+
+/// function that computes kinetic energy
+void ParticleMeanEnergy::ComputeDiags (int step)
+{
+
+    /// Judge if the diags should be done
+    if ( (step+1) % m_freq != 0 ) { return; }
+
+    /// get WarpX class object
+    auto & warpx = WarpX::GetInstance();
+
+    /// get MultiParticleContainer class object
+    auto & mypc = warpx.GetPartContainer();
+
+    /// get number of species (int)
+    auto species_number = mypc.nSpecies();
+
+    /// resize data array
+    /// the extra one is for total energy
+    m_data.resize(species_number+1,0.0);
+
+    /// get species names (std::vector<std::string>)
+    auto species_names = mypc.GetSpeciesNames();
+
+    /// get number of level (int)
+    auto level_number = warpx.finestLevel();
+
+    /// speed of light squared
+    auto c2 = PhysConst::c * PhysConst::c;
+
+    /// declare total kinetic energy variable
+    amrex::Real EKtot = 0.0;
+
+    /// total number of particles;
+    long nptot = 0;
+
+    /// loop over species
+    for (int i_s = 0; i_s < species_number; ++i_s)
+    {
+        /// get WarpXParticleContainer class object
+        auto & myspc = mypc.GetParticleContainer(i_s);
+
+        /// get mass (amrex:Real)
+        auto m = myspc.getMass();
+
+        /// declare kinetic energy variable
+        amrex::Real EK = 0.0;
+
+        /// number of particles;
+        long np = 0;
+
+        /// loop over refinement levels
+        for (int lev = 0; lev <= level_number; ++lev)
+        {
+
+            #pragma omp parallel reduction(+:EK, np)
+            /// Loop over boxes
+            for (WarpXParIter pti(myspc, lev); pti.isValid(); ++pti)
+            {
+
+                /// get particle momentum arrays
+                auto & px = pti.GetAttribs(PIdx::ux);
+                auto & py = pti.GetAttribs(PIdx::uy);
+                auto & pz = pti.GetAttribs(PIdx::uz);
+
+                /// sum total number of particles
+                np += pti.numParticles();
+
+                /// loop over particles
+                for (long i = 0; i < px.size(); i++)
```

Comment:
This will not run on GPU.
@atmyers @WeiqunZhang does AMReX have reduction functions on GPU (e.g., average a quantity over all particles in a box)?

Reply:
Yes, check out `Src/Particle/AMReX_ParticleReduceH`.

Adding docstrings for these is on my list.